### PR TITLE
Add more EntityDismountEvent API

### LINKED ISSUE #6183: add reason to entitydismountevent
### Is your feature request related to a problem?

I want to prevent dismounting when entering water

### Describe the solution you'd like.

A entitydismountevent reason, eg Reason.IN_WATER

### Describe alternatives you've considered.

Forking paper

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### Other

_No response_

### PR DESCRIPTION
**Pull request by LemonCaramel**
Closes #6183

I don't live in an English-speaking country, so the API description is not appropriate.

Feedback is always welcome.

### CODE REVIEW

File: patches/api/0330-Add-more-EntityDismountEvent-API.patch [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
++         */
++        DISCONNECT,
++        /**
++         * Mounted on a new vehicle while mount the vehicle
++         */
++        NEW_MOUNT,
++        /**
++         * Passenger pressed the unmount key
++         */
++        SHIFT_KEY,
```
Comment (lynxplay) [inline patches/api/0330-Add-more-EntityDismountEvent-API.patch]: I would personally not call this `SHIFT_KEY` concerning that people can really map their unmount key to whatever they want.
`UNMOUNT` would probably fit better and is a lot more general.

File: patches/api/0330-Add-more-EntityDismountEvent-API.patch [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
++        VANILLA_AI
++    }
++    // Paper end - Add EntityDismountEvent Reason
+ 
+     public EntityDismountEvent(@NotNull Entity what, @NotNull Entity dismounted) {
+         // Paper start
+@@ -23,10 +123,17 @@ public class EntityDismountEvent extends EntityEvent implements Cancellable {
+ 
+ 
+     public EntityDismountEvent(@NotNull Entity what, @NotNull Entity dismounted, boolean isCancellable) {
```
Comment (Machine-Maker) [inline patches/api/0330-Add-more-EntityDismountEvent-API.patch]: Deprecate this ctor (assuming its not used in the server anywhere, idk)

**Review by Machine-Maker — APPROVED** [review]

**Comment by LemonCaramel** [comment]
hello bot

**Review by electronicboy — APPROVED** [review]
I think that the names of some of the reasons are potentially iffy, but, unless somebody has better name suggestions, lgtm

**Pull request closed without merging**